test/index.js

```diff
@@ -3877,3 +3877,140 @@ suite ('interoperability', () => {
+         Left ({'error': 'WrongValue', 'name': 'foo', 'type': 'String', 'value': null}),
+       ]);
+
+    eq ($.validate (FooBar) ({foo: 'blue', bar: 42}))
+       ([Right ({foo: 'blue', bar: 42})]);
+
+  });
+
+  test ('Custom Type', () => {
+
+    //    $DateIso :: NullaryType
+    const $DateIso = (
+      $.NullaryType ('DateIso')
+                    ('https://www.w3.org/QA/Tips/iso-date')
+                    ([$.String])
+                    (x => /^\d{4}-(?:0[1-9]|1[0-2])-(?:0[1-9]|[1-2]\d|3[0-1])$/.test (x))
+    );
+
+    const model1 = $.RecordType ({
+      date: $DateIso,
+    });
+
+    const model2 = $.RecordType ({
+      date: $.NonEmpty ($DateIso),
+      bool: $.Boolean,
+    });
+
+    eq ($.validate (model1) ({date: '2020-04-10'}))
+       ([Right ({date: '2020-04-10'})]);
+
+    eq ($.validate (model1) ({date: '2020-04-100'}))
+       ([
+         Left ({'error': 'WrongValue', 'name': '$$', 'type': 'RECORD', 'value': {'date': '2020-04-100'}}),
+         Left ({'error': 'WrongValue', 'name': 'date', 'type': 'DateIso', 'value': '2020-04-100'}),
+       ]);
+
+    eq ($.validate (model2) (undefined))
+       ([
+         Left ({'error': 'WrongValue', 'name': '$$', 'type': 'RECORD', 'value': undefined}),
+         Left ({'error': 'MissingValue', 'name': 'date', 'type': 'RECORD', 'value': undefined}),
+         Left ({'error': 'MissingValue', 'name': 'bool', 'type': 'RECORD', 'value': undefined}),
+       ]);
+
+    eq ($.validate (model2) ({bool: 'foobar', date: '2020-04-100'}))
+       ([
+        Left ({'error': 'WrongValue', 'name': '$$', 'type': 'RECORD', 'value': {'bool': 'foobar', 'date': '2020-04-100'}}),
+        Left ({'error': 'WrongValue', 'name': 'date', 'type': 'DateIso', 'value': '2020-04-100'}),
+        Left ({'error': 'WrongValue', 'name': 'bool', 'type': 'Boolean', 'value': 'foobar'}),
+       ]);
+
+    eq ($.validate (model2) ({date: '2020-04-10', bool: 'foobar'}))
+       ([
+         Left ({'error': 'WrongValue', 'name': '$$', 'type': 'RECORD', 'value': {'bool': 'foobar', 'date': '2020-04-10'}}),
+         Left ({'error': 'WrongValue', 'name': 'bool', 'type': 'Boolean', 'value': 'foobar'}),
+       ]);
+
+    eq ($.validate (model2) ({date: '2020-04-100', bool: true}))
+       ([
+         Left ({'error': 'WrongValue', 'name': '$$', 'type': 'RECORD', 'value': {'bool': true, 'date': '2020-04-100'}}),
+         Left ({'error': 'WrongValue', 'name': 'date', 'type': 'DateIso', 'value': '2020-04-100'}),
+       ]);
+
+    eq ($.validate (model2) ({date: [], bool: false}))
+       ([
+        Left ({'error': 'WrongValue', 'name': '$$', 'type': 'RECORD', 'value': {'bool': false, 'date': []}}),
+        Left ({'error': 'WrongValue', 'name': 'date', 'type': 'NonEmpty', 'value': []}),
+       ]);
+
+    eq ($.validate (model2) ({bool: false, date: '2020-04-10'}))
+       ([Right ({date: '2020-04-10', bool: false})]);
```

Comment:
`Array (Either ValidationError a)` permits invalid combinations such as `[Left (e), Right (x)]` and `[Right (x), Right (x), Right (x)]`. We could make such invalid combinations unrepresentable by changing the nesting of the types:

```haskell
Either (Array ValidationError) a
```

What do you think of this option?

Reply:
I'm wondering if there is a use-case for having `Right (x)`s when `$$` fails. Basically removing [step 4](https://github.com/sanctuary-js/sanctuary-def/pull/308/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R1592-R1595).

```js
/* [
Left ({"error": "WrongValue", "name": "$$", "type": "RECORD", "value": {"bool": false}}),
Left ({"error": "MissingValue", "name": "date", "type": "NonEmpty", "value": undefined}),
Right ({"name": "bool", "result": Right (false), "type": Boolean, "value": false})
] */
$.validate (model2) ({bool: false});
```

But if there is not then I think your suggestion is great!

I still have 3 questions though.

1. How do I define `ValidationError`? As a POJO (as in this PR), via `def` or using daggy?
2. is `ValidationError` a sum type of `MissingValue` and `WrongValue`? _See below †_
3. What is the possible values contained in `p.result.value.propPath`? Can it be hard-coded as `p.type.types.$1`?

† I'm not super familiar with sum types but as I understand, the encoding would be something like below, for returning an array of `Either`s:

```reason
type Error   = Left (ValidationError);
type Success = Right (x);

type result  = Error (String, String, String, x)
             | Success (x);

let validate = (Type, x) => {
	/* validation logic */
	[result, result, result, ...result];
};
```

Reply:
> I'm wondering if there is a use-case for having `Right (x)`s when `$$` fails.

In my experience it's rare that one is interested in what *succeeded* when validation fails.

Let's change the nesting. :)